File: python/57-Insert-Interval.py
from typing import List

# region Collect Left & Right
# Space O(n)
# Time O(n)


class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        curStart, curEnd = newInterval[0], newInterval[1]

        left, right = [], []

        for i in intervals:
            if i[1] < curStart:
                left.append(i)
            elif i[0] > curEnd:
                right.append(i)
            else:
                curStart = min(curStart, i[0])
                curEnd = max(curEnd, i[1])
        return left + [[curStart, curEnd]] + right


# endregion

# region  Hacky Linear Search -- Can be improved
# Space O(N)
# Time O(N)


class Solution:
    def insert(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        if not intervals:
            return [newInterval]

        retVal = []

        curStart, curEnd = newInterval[0], newInterval[1]

        for i in range(len(intervals)):
            interval = intervals[i]
            # Old interval ends before curStart
            if interval[1] < curStart:
                retVal.append(interval)
                continue
            # Cur interval start overlaps with old
            if (
                interval[0] <= curStart
                and curStart <= interval[1]
                or (curEnd >= interval[0] and curEnd <= interval[1])
            ):
                curStart = min(interval[0], curStart)
                curEnd = max(curEnd, interval[1])
            if interval[0] > curEnd:
                retVal.append([curStart, curEnd])
                retVal.extend(intervals[i:])
                return retVal
        retVal.append([curStart, curEnd])
        return retVal


# endregion

File: c#/57-Insert-Interval.cs
// Time O(n)
// Space O(1)
public class Solution
{
    public int[][] Insert(int[][] intervals, int[] newInterval)
    {
        // Iterate through the intervals
        // If interval starts after new one, add new interval and add remaining intervals then return
        // If interval ends before new interval, add interval
        // Otherwise, edit newInterval start & end to be min & max of newInterval and current interval values
        // If we get out of the for loop, no intervals were after new one so we need to add newInterval

        List<int[]> retVal = new();

        for (int i = 0; i < intervals.Length; i++)
        {
            var interval = intervals[i];
            if (interval[0] > newInterval[1])
            {
                retVal.Add(newInterval);
                retVal.AddRange(intervals.Skip(i));
                return retVal.ToArray();
            }
            else if (interval[1] < newInterval[0])
            {
                retVal.Add(interval);
            }
            else
            {
                newInterval[0] = Math.Min(interval[0], newInterval[0]);
                newInterval[1] = Math.Max(interval[1], newInterval[1]);
            }
        }

        retVal.Add(newInterval);

        return retVal.ToArray();
    }
}
